### DISCUSSION TITLE
Trial GitHub Actions vs Cirrus CI

### DISCUSSION THREAD

**Discussion by trexfeathers** [SciTools/iris#4193] [Ideas]
## 📰 Custom Issue

@ocefpaf has correctly pointed out (https://github.com/SciTools/cf-units/pull/175, https://github.com/SciTools/nc-time-axis/pull/66) that GHA is the most popular CI implementation for scientific Python projects. Given that barriers to involvement are a big factor in the health of a project: we need a **really good** reason to go against the crowd as we do now using Cirrus CI.

I believe our reasons for using Cirrus are to minimise the runtime for a full CI run. **We should therefore do a trial to see if it possible to get similar/faster runtime using GHA instead.** I'm told that GHA also has caching solutions like Cirrus does, and the minutes are free for public repositories so there should be no trouble with parallelism.

**Comment by trexfeathers** [comment]
I probably shouldn't have been so quick to shoot down @jamesp's idea on https://github.com/SciTools/iris-grib/issues/242!

**Reply by jamesp** [reply]
@trexfeathers it wasn't the right solution at the time, it might be now... either way all good!

**Comment by rcomer** [comment]
Cartopy went to GHA in December https://github.com/SciTools/cartopy/pull/1691

**Comment by bjlittle** [comment]
We could easily move to GH actions for CI. I'm certainly not wedded to `cirrus-ci` - it certainly dug us out of a hole quickly when `travis-ci` was causing us major issues. Also, what's seriously going against `cirrus-ci` is having to pay credits for reasonable resource 🧐

I've seen several major projects migrating and using it, and I'm bootstrapping [GeoVista](https://github.com/bjlittle/geovista) with GH actions instead of `cirrus-ci`. It has an interesting use case, as it also needs Mesa support for PyVista, and I want to target `linux`, `osx` and eventually `windows` (but `iris` needs a bit of work for that) - anyways, all of that is easy to do and supported for GH actions.

I don't believe this will be a difficult thing to do, and the interface with `nox` as it stands, shouldn't need any, or very little, change.

**Reply by bjlittle** [reply]
Also, if we're rebooting CI, for me idealistically this would include:
- testing on `ubuntu-latest`, `osx-latest`, `windows-latest`
- testing both `conda` and `pypi` packaging
- pushing a slim / distro-less `docker` image to `dockerhub` for stable releases

**Reply by bjlittle** [reply]
We should openly discuss `nox` vs `tox`

**Reply by bjlittle** [reply]
Also see https://github.com/SciTools/nc-time-axis/pull/57

**Reply by jamesp** [reply]
Agree it's worth considering tox vs nox.  See https://github.com/tox-dev/tox-conda which has support for `@EXPLICIT` conda-lock files

**Comment by trexfeathers** [comment]
It sounds like there are [some exciting possibilities for parallelisation with GHA+pytest](https://blog.jerrycodes.com/pytest-split-and-github-actions/), so that might be another argument for switching.

**Comment by tkknight** [comment]
Worth considering using GH actions to build the docs also on each pull request so that it is viewable by the reviewers.  

Discussion: https://github.com/SciTools/iris/discussions/4398

**Comment by pp-mo** [comment]
#4503 probably addresses
